strictdoc/core/file_traceability_index.py

```diff
@@ -433,16 +424,39 @@ def marker_comparator(marker):
                 traceability_info_.ng_lines_total, coverage
             )
 
-            for markers_ in traceability_info_.ng_map_reqs_to_markers.values():
+            for (
+                req_uid,
+                markers,
+            ) in traceability_info_.ng_map_reqs_to_markers.items():
 
                 def marker_comparator(marker):
                     return marker.ng_range_line_begin, marker.ng_range_line_end
 
-                markers_.sort(key=marker_comparator)
+                markers.sort(key=marker_comparator)
+
+                # validate here, SDocNode.relations doesn't track marker roles
+                node = traceability_index.get_node_by_uid(req_uid)
+                assert node.document and node.document.grammar
+                grammar_element = node.document.grammar.elements_by_type[
+                    node.node_type
+                ]
+                for marker in markers:
+                    if not grammar_element.has_relation_type_role(
+                        relation_type="File",
+                        relation_role=marker.role,
+                    ):
+                        raise StrictDocSemanticError.invalid_marker_role(
+                            node=node,
+                            marker=marker,
+                            path_to_src_file=path,
+                        )
+
+        def path_with_role_comparator(path_with_role):
+            return path_with_role[0]
 
         # Sort by keys alphabetically.
-        for paths_ in self.map_reqs_uids_to_paths.values():
-            paths_.sort()
+        for paths_with_role in self.map_reqs_uids_to_paths_with_role.values():
+            paths_with_role.sort(key=path_with_role_comparator)
 
     def create_requirement(
```

Comment:
Note that the create_requirement was renamed to `create_requirement_with_forward_source_links` and it now looks like this:

```
    def create_requirement_with_forward_source_links(
        self, requirement: SDocNode
    ) -> None:
        self.requirements_with_forward_links.add(requirement)
```

all the logic was moved to be a separate step in `validate_in_resolve`.

Reply:
All the lines should be changed in `validate_in_resolve`, so there should be not much room for surprises.